acme/acme/messages.py

```diff
@@ -190,25 +193,28 @@ class Identifier(jose.JSONObjectWithFields):
     """ACME identifier.
 
     :ivar IdentifierType typ:
-    :ivar unicode value:
+    :ivar str value:
 
     """
-    typ = jose.Field('type', decoder=IdentifierType.from_json)
-    value = jose.Field('value')
+    typ: IdentifierType = jose.field('type', decoder=IdentifierType.from_json)
+    value: str = jose.field('value')
+
+
+R = TypeVar('R', bound=ResourceMixin)
 
 
 class Directory(jose.JSONDeSerializable):
     """Directory."""
 
-    _REGISTERED_TYPES: Dict[str, Type['Directory']] = {}
+    _REGISTERED_TYPES: Dict[str, Type[ResourceMixin]] = {}
```

Comment:
I don't love the design of using `ResourceMixin` here and as the bound of the `TypeVar` above. My reasons for this are the `ResourceMixin` is about the `resource` attribute while the `Directory` cares about the `resource_type` attribute. To support this design, we had to add `resource_type` to `ResourceMixin` as well as make `NewOrder` inherit from `ResourceMixin` which is not needed because a `NewOrder` should never have a `resource` attribute.

Is there something else we can do here, my first thought would be to use [protocol](https://mypy.readthedocs.io/en/stable/protocols.html#simple-user-defined-protocols). To get `typing_extensions` in `acme`, we could take an approach like the one we did at https://github.com/certbot/certbot/commit/0f9f902b6eb32bab609922cdcc78649d78c7b7c0.

What do you think?

Reply:
I totally agree. I was hesitant initially to use protocols here, but the context you add on `ResourceMixin` convinced me that it is a better move.